proxy_ring_buffer.go

```diff
@@ -0,0 +1,278 @@
+}
+
+// TailSequence returns the sequence of the tail. The tail is the side of the ringbuffer where the items are added to.
+// The initial value of the tail is -1.
+func (rb *Ringbuffer) TailSequence(ctx context.Context) (int64, error) {
+	request := codec.EncodeRingbufferTailSequenceRequest(rb.name)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferTailSequenceResponse(response), nil
+}
+
+// HeadSequence returns the sequence of the head. The head is the side of the ringbuffer where the oldest items in the ringbuffer
+// are found. If the RingBuffer is empty, the head will be one more than the tail.
+// The initial value of the head is 0 (1 more than tail).
+func (rb *Ringbuffer) HeadSequence(ctx context.Context) (int64, error) {
+	request := codec.EncodeRingbufferHeadSequenceRequest(rb.name)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferHeadSequenceResponse(response), nil
+}
+
+// RemainingCapacity
+// Returns the remaining capacity of the Ringbuffer. The returned value could be stale as soon as it is returned.
+// If ttl is not set, the remaining capacity will always be the capacity.
+func (rb *Ringbuffer) RemainingCapacity(ctx context.Context) (int64, error) {
+	request := codec.EncodeRingbufferRemainingCapacityRequest(rb.name)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferRemainingCapacityResponse(response), nil
+}
+
+// ReadMany
+// Reads a batch of items from the Ringbuffer. If the number of available items after the first read item is smaller
+// than the maxCount, these items are returned. So it could be the number of items read is smaller than the maxCount.
+// If there are fewer items available than minCount, then this call blacks. Reading a batch of items is likely to
+// perform better because less overhead is involved. A filter can be provided to only select items that need to be read.
+// If the filter is null, all items are read. If the filter is not null, only items where the filter function returns
+// true are returned. Using filters is a good way to prevent getting items that are of no value to the receiver.
+// This reduces the amount of IO and the number of operations being executed, and can result in a significant performance improvement.
+func (rb *Ringbuffer) ReadMany(ctx context.Context, startSequence int64, minCount int32, maxCount int32, filter interface{}) (ReadResultSet, error) {
+	if filter != nil {
+		return ReadResultSet{}, errors.New("filter functions are not yet supported in th Go client, please use 'nil' for now")
+	}
+	request := codec.EncodeRingbufferReadManyRequest(rb.name, startSequence, minCount, maxCount, nil)
```

Comment:
@srknzl 

I did implement the defensive checks for "if xxx <0 ...".
See 4ed59c73d180f7f1a0c71272d8fe2fb047b2658a

I hesitate to implement the checks about capacity because this is something the server will do anyway.
if we implement this client-side, this requires extra lookups (yes cached, I know).
When there are any plans of e.g. dynamic resizing, these checks on client-side are difficult anyhow.
What I'm trying to say: I do propose to not implement them, as they are beyond the API's data contract and so should be implemented in one place only.

Would you consider agreeing, leaving in just the non-negative tests.

Reply:
I don't know why Java client has it but can you add it for now. We can discuss why we have it in our team internally. If we will remove it, it should be removed from Java first.

Reply:
Done
d15128b24b9e59278ddd608920811529820f3a93